api/pkg/user/service/service.go

```diff
@@ -0,0 +1,201 @@
+		if err != nil {
+			http.Error(res, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		scheme := &JWTScheme{
+			Name:   jwt,
+			Scopes: []string{"rating:read", "rating:write", "agent:create", "catalog:refresh", "config:refresh", "refresh:token"},
+		}
+
+		err = ValidateScopes(claims, scheme)
+		if err != nil {
+			http.Error(res, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		userID, ok := claims["id"].(float64)
+		if !ok {
+			http.Error(res, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		// Set the userId in the headers
+		req.Header.Set("UserId", fmt.Sprintf("%v", userID))
+
+		handler.ServeHTTP(res, req)
+	}
+}
+
+// ValidateScopes takes user scopes and checks if it has the scope which
+// is required for accessing the api
+func ValidateScopes(claims jwt.MapClaims, scheme *JWTScheme) error {
+
+	if claims["scopes"] == nil {
+		return fmt.Errorf("invalid scopes")
+	}
+
+	scopes, ok := claims["scopes"].([]interface{})
+	if !ok {
+		return fmt.Errorf("invalid scopes")
+	}
+
+	scopesInToken := make([]string, len(scopes))
+	for _, scp := range scopes {
+		scopesInToken = append(scopesInToken, scp.(string))
+	}
+	if err := scheme.Validate(scopesInToken); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Validate returns a non-nil error if scopes does not contain all of
+// JWT scheme's required scopes.
+func (s *JWTScheme) Validate(scopes []string) error {
+	return validateScopes(s.RequiredScopes, scopes)
+}
+
+func validateScopes(expected, actual []string) error {
+	var missing []string
+	for _, r := range expected {
+		found := false
+		for _, s := range actual {
+			if s == r {
+				found = true
+				break
+			}
+		}
+		if !found {
+			missing = append(missing, r)
+		}
+	}
+	if len(missing) == 0 {
+		return nil
+	}
+	return fmt.Errorf("missing scopes: %s", strings.Join(missing, ", "))
+}
+
+func (r *request) User(id int) (*model.User, error) {
+
+	var user model.User
+	if err := r.db.First(&user, id).Error; err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			r.log.Warnf("user not found for token: %s", err.Error())
+			return nil, err
+		}
+
+		r.log.Errorf("error when looking up user. err: %s", err.Error())
+		return nil, err
+	}
+
+	return &user, nil
+}
+
+func (s *UserService) newRequest(user *model.User) *request {
+	return &request{
+		db:            s.DB(context.Background()),
+		log:           s.Logger(context.Background()),
+		user:          user,
+		defaultScopes: s.api.Data().Default.Scopes,
+		jwtConfig:     s.api.JWTConfig(),
+	}
+}
+
+func (s *UserService) validateRefreshToken(id int, token string) (*model.User, error) {
+
+	r := request{
+		db:            s.DB(context.Background()),
+		log:           s.Logger(context.Background()),
+		defaultScopes: s.api.Data().Default.Scopes,
+		jwtConfig:     s.api.JWTConfig(),
+	}
+
+	user, err := r.User(id)
+	if err != nil {
+		r.log.Error(err)
+		return nil, err
+	}
+
+	if user.RefreshTokenChecksum != createChecksum(token) {
```

Comment:
token contains Header along with token value and needs to split the token value

Reply:
You mean `Bearer + token` ??

Reply:
yes